iothub/device/samples/PnpDeviceSamples/TemperatureController/Program.cs

```diff
@@ -0,0 +1,327 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using PnpHelpers;
+
+namespace TemperatureController
+{
+    internal enum StatusCode
+    {
+        Completed = 200,
+        InProgress = 202,
+        NotFound = 404
+    }
+
+    public class Program
+    {
+        private const string ModelId = "dtmi:com:example:TemperatureController;1";
+        private const string Thermostat1 = "thermostat1";
+        private const string Thermostat2 = "thermostat2";
+        private const string SerialNumber = "SR-123456";
+
+        private static readonly string s_deviceConnectionString = Environment.GetEnvironmentVariable("IOTHUB_DEVICE_CONNECTION_STRING");
+        private static readonly Random s_random = new Random();
+
+        private static DeviceClient s_deviceClient;
+        private static ILogger s_logger;
+
+        // A dictionary to hold all desired property change callbacks that this pnp device should be able to handle.
+        // The key for this dictionary is the componentName.
+        private static readonly IDictionary<string, DesiredPropertyUpdateCallback> s_desiredPropertyUpdateCallbacks = new Dictionary<string, DesiredPropertyUpdateCallback>();
+
+        // Dictionary to hold the temperature updates sent over each "Thermostat" component.
+        private static readonly Dictionary<string, Dictionary<DateTimeOffset, double>> s_temperatureReadings = new Dictionary<string, Dictionary<DateTimeOffset, double>>();
+
```

Comment:
This is very strange to me. We want a device to hold all "temperature updates" info which might be restricted by memory and other resource constraints?

Reply:
Why couldn't we just push it as a property and let service client hold it?

Reply:
We need a container to hold all temperature updates, in order to calculate the avg temp since last reboot -> this is required for "getMaxMinReport" command.

Reply:
I can modify this to hold a running total instead, and calculate the avg from that; but do we need something like that for a sample?

Reply:
On second thought, even a running total for average wouldn't work, since the report needs to be generated from a point in time. So we need to store the historical data somewhere. 
We cannot send it to service since reported properties will get overwritten with each update, and sending it via telemetry will require integration with a storage related service.

Reply:
Yeah that is my concern. That storing seems to be inbuilt into Model which I don't think is a good use case. Small devices wouldn't want to store rather won't be able to store. They need to able to send data as soon as they have it ready

Reply:
Yes, in those scenarios they would need to leverage the storage capabilities of a different service.  I will add that as a comment.

Reply:
If the reboot doesn't happen often (which is usually true in case of IOT devices, sample demonstrates to store all the data generated which is not what a typical PnP user would want to do.